### PR TITLE
TT-5144 MDCB SQL storage

### PR DESCRIPTION
**Pull request by tbuchaillot**
Adding a note in `Setup Master Data Centre` for SQL storage configuration.

Updating MDCB configuration options

### CODE REVIEW

**Review by marksou — APPROVED** [review]
LGTM

**Comment by sredxny** [comment]
Updated the config options for MDCB. I removed any reference to the synchronizer, as this feature will be published once Tyk v4.1 is released

**Review by sredxny — APPROVED** [review]
lgtm

**Comment by sredxny** [comment]
@marksou I think this is ready to be published. Let us know if you need anything else. This PR complements the release of MDCB v2.0 done this week

**Comment by sredxny** [comment]
@andyo-tyk take a look please

File: tyk-docs/content/shared/mdcb-config.md [hunk 1]
```diff
@@ -450,43 +450,3 @@ EV: <b>TYK_MDCB_ENABLEKEYLOGGING</b><br />
 Type: `bool`<br />
 
 EnableKeyLogging prints the unhashed keys without obfuscate them in the logs
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:452]: Please change this to:
EnableKeyLogging prints the unhashed keys (without obfuscating them) to the logs

**Review by andyo-tyk — CHANGES_REQUESTED** [review]
Please improve this wording, thanks.

File: tyk-docs/content/shared/mdcb-config.md [hunk 2]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_TYPE</b><br />
+Type: `DBType`<br />
+
+Determines the storage type. It could be `mongo`, `postgres` or `sqlite`. By default, the value is `mongo`.
+
+### analytics.connection_string
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_CONNECTIONSTRING</b><br />
+Type: `string`<br />
+
+This is used to configure the conenction string for the storage.
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:164]: Typo: please correct to "connection".

Do we define what a conection string is, somewhere? I note that it's referenced below as well in the description of TYK_MDCB_ANALYTICSCONFIG_MONGOURL - how do these two work together?

File: tyk-docs/content/shared/mdcb-config.md [hunk 3]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_MYSQL_DONTSUPPORTRENAMECOLUMN</b><br />
+Type: `bool`<br />
+
+`change` when rename column, rename column not supported before MySQL 8, MariaDB
+
+### analytics.mysql.skip_initialize_with_version
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_MYSQL_SKIPINITIALIZEWITHVERSION</b><br />
+Type: `bool`<br />
+
+auto configure based on currently MySQL version
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:212]: Need to capitalise the first word, and I think there's also a typo. Perhaps should be:

"Auto configure based on current MySQL version"

....but auto configure what?

File: tyk-docs/content/shared/mdcb-config.md [hunk 4]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_MYSQL_DISABLEDATETIMEPRECISION</b><br />
+Type: `bool`<br />
+
+disable datetime precision, which not supported before MySQL 5.6
+
+### analytics.mysql.dont_support_rename_index
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_MYSQL_DONTSUPPORTRENAMEINDEX</b><br />
+Type: `bool`<br />
+
+drop & create when rename index, rename index not supported before MySQL 5.7, MariaDB
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:200]: This needs to be capitalised, but is also not clear (to me).

File: tyk-docs/content/shared/mdcb-config.md [hunk 5]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_MYSQL_DEFAULTSTRINGSIZE</b><br />
+Type: `uint`<br />
+
+default size for string fields. By default set to: 256
+
+### analytics.mysql.disable_datetime_precision
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_MYSQL_DISABLEDATETIMEPRECISION</b><br />
+Type: `bool`<br />
+
+disable datetime precision, which not supported before MySQL 5.6
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:194]: Needs to be capitalised, but also some rewording suggested:

"Disable datetime precision; this is required for databases prior to MySQL 5.6 (which do not support datetime precision)."

File: tyk-docs/content/shared/mdcb-config.md [hunk 6]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_BATCHSIZE</b><br />
+Type: `int`<br />
+
+Max Batch size for SQL Analytics
+
+### analytics.postgres.prefer_simple_protocol
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_POSTGRES_PREFERSIMPLEPROTOCOL</b><br />
+Type: `bool`<br />
+
+disables implicit prepared statement usage
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:182]: First word needs to be capitalised

File: tyk-docs/content/shared/mdcb-config.md [hunk 7]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
+
+### analytics.mysql.default_string_size
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_MYSQL_DEFAULTSTRINGSIZE</b><br />
+Type: `uint`<br />
+
+default size for string fields. By default set to: 256
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:188]: Perhaps reword to something like this:

"Set a default size for string fields; if this is not set then it will default to 256 characters."

File: tyk-docs/content/shared/mdcb-config.md [hunk 8]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_MYSQL_DONTSUPPORTRENAMEINDEX</b><br />
+Type: `bool`<br />
+
+drop & create when rename index, rename index not supported before MySQL 5.7, MariaDB
+
+### analytics.mysql.dont_support_rename_column
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_MYSQL_DONTSUPPORTRENAMECOLUMN</b><br />
+Type: `bool`<br />
+
+`change` when rename column, rename column not supported before MySQL 8, MariaDB
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:206]: Again, this isn't completely clear what it does - is it just provided for users trying to SQL versions prior to MySQL8 and MariaDB?

File: tyk-docs/content/shared/mdcb-config.md [hunk 9]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
 Set the consistency mode for the session, it defaults to `Strong`. The valid values are:
 * eventual
 monotonic
 
 ### analytics.mongo_batch_size
-EV: **TYK_MDCB_ANALYTICSCONFIG_MONGOBATCHSIZE**<br />
+EV: <b>TYK_MDCB_ANALYTICSCONFIG_MONGOBATCHSIZE</b><br />
 Type: `int`<br />
 
 Sets the batch size for mongo results.
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:262]: Mongo should be capitalised

File: tyk-docs/content/shared/mdcb-config.md [hunk 10]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
 Type: `bool`<br />
 
 Instead of saving all the analytics in one key, this will enable to save the analytics in multiple keys. It's specially useful when you are using Redis cluster. This will work only if `forward_analytics_to_pump` is true and tyk-pump is v1.2.1+ .
 
 ### dont_store_selective
-EV: **TYK_MDCB_DONTSTORESELECTIVE**<br />
+EV: <b>TYK_MDCB_DONTSTORESELECTIVE</b><br />
 Type: `bool`<br />
 
 set to true if you don't want to store selective analytics
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:292]: First word needs to be capitalised

File: tyk-docs/content/shared/mdcb-config.md (hunk 2 above)
Comment (sredxny) [reply tyk-docs/content/shared/mdcb-config.md:164]: @andyo-tyk in database terminology string connection is a very common word, and people configuring connection to databases know what it is

File: tyk-docs/content/shared/mdcb-config.md [hunk 11]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
 
 ### dont_store_aggregate
-EV: **TYK_MDCB_DONTSTOREAGGREGATES**<br />
+EV: <b>TYK_MDCB_DONTSTOREAGGREGATES</b><br />
 Type: `bool`<br />
 
 Set to true to don't store aggregate analytics
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:298]: So that it matches the previous feature, reword to:

"Set to true if you don't want to store aggregate analytics."

File: tyk-docs/content/shared/mdcb-config.md [hunk 12]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
 Type: `bool`<br />
 
 Currently, analytics for an endpoint is stored only if Track Endpoint plugin is enabled on that endpoint. If `track_all_paths` is enabled, it will store analytics for all the endpoints, irrespective of Track Endpoint plugin.
 
 ### store_analytics_per_minute
-EV: **TYK_MDCB_STOREANALYTICSPERMINUTE**<br />
+EV: <b>TYK_MDCB_STOREANALYTICSPERMINUTE</b><br />
 Type: `bool`<br />
 
 Enable to generate aggregated per minute. By default it will generate aggregate data per hour. If this option is enabled, aggregate data will be generated per minute.
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:328]: Perhaps simplify this to:

"Set this to true to aggregate the analytics per minute; the default behviour otherwise is that data will be aggregated per hour."

File: tyk-docs/content/shared/mdcb-config.md [hunk 13]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
 
+### ignore_tag_prefix_list
+EV: <b>TYK_MDCB_IGNORETAGPREFIXLIST</b><br />
+Type: `[]string`<br />
+
+if set to true then it will not store analytics for tags having prefix specified in the list. **Note**: Prefix “key-” is added in the list by default. This tag is added by gateway for keys.
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:334]: The first word needs to be capitalised but perhaps reword this to:

"Use this field to define a list of tags for which you do not want to store analytics; any analytics that are generated and tagged by the gateway with a prefix in this list will be discarded.
Note that the prefix "key-" is automatically included in this list and is used by the gateway for keys."

**Comment by sredxny** [comment]
@andyo-tyk I will make the changes. It requires changes in tyk-sink repo and then regenerate again the documentation.

File: tyk-docs/content/shared/mdcb-config.md [hunk 14]
```diff
@@ -61,222 +59,288 @@ Allows MDCB to use Mutual TLS. This requires that `server_options.use_ssl` is se
...
+
 ### threshold_len_tag_list
-EV: **TYK_MDCB_THRESHOLDLENTAGLIST**<br />
+EV: <b>TYK_MDCB_THRESHOLDLENTAGLIST</b><br />
 Type: `int`<br />
 
  If number of tags in a document grows beyond `threshold_len_tag_list`, pump will throw a warning, it works for mongo aggregate pump. The warning will print top 5 common tag prefix. Default value is 1000. To disable alerts set it to -1.
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:340]: Does this ONLY work for the Mongo Aggregate pump?

If "yes" perhaps reword to:
"When using the Mongo Aggregate Pump, configure this value for the Pump to throw a warning if the number of tags in a document grows beyond this limit. The warning will print top 5 common tag prefix. If this is not configured, a default value of 1000 will be used. To disable alerts set it to -1."

File: tyk-docs/content/shared/mdcb-config.md [hunk 15]
```diff
@@ -285,99 +349,144 @@ Set it to true if you are using a separated analytic storage in the master gatew
...
 Type: `string`<br />
 
 It defines the sentinel master name
 
 ### analytics_storage.sentinel_password
-EV: **TYK_MDCB_ANALYTICSSTORAGE_SENTINELPASSWORD**<br />
+EV: <b>TYK_MDCB_ANALYTICSSTORAGE_SENTINELPASSWORD</b><br />
 Type: `string`<br />
 
 If set, redis sentinel will authenticate using this password.
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:379]: Capitalise Redis

File: tyk-docs/content/shared/mdcb-config.md [hunk 16]
```diff
@@ -285,99 +349,144 @@ Set it to true if you are using a separated analytic storage in the master gatew
...
 
 ### analytics_storage.username
-EV: **TYK_MDCB_ANALYTICSSTORAGE_USERNAME**<br />
+EV: <b>TYK_MDCB_ANALYTICSSTORAGE_USERNAME</b><br />
 Type: `string`<br />
 
 If set, a redis connection will be established with this user. If not set then it will defaults to the default redis user
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md:385]: Capitalise Redis

File: tyk-docs/content/shared/mdcb-config.md [hunk 17]
```diff
@@ -285,99 +349,144 @@ Set it to true if you are using a separated analytic storage in the master gatew
...
 You can now set a logging level (log_level). The following levels can be set: debug, info, warn, error.
 If not set or left empty, it will default to `info`.
 
+### enable_key_logging
+EV: <b>TYK_MDCB_ENABLEKEYLOGGING</b><br />
+Type: `bool`<br />
+
+EnableKeyLogging prints the unhashed keys without obfuscate them in the logs
+
+### sync_worker_config
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md]: If this section relates to the synchroniser, I think we need to remove it until Tyk 4.1 is released?

File: tyk-docs/content/shared/mdcb-config.md [hunk 18]
```diff
@@ -285,99 +349,144 @@ Set it to true if you are using a separated analytic storage in the master gatew
...
+EV: <b>TYK_MDCB_SYNCWORKER_HASHKEYS</b><br />
+Type: `bool`<br />
+
+Allows the worker to synchronize hashed API keys. Set this to true if `hash_keys` is true in dashboard and gateway configuration.
+
+### sync_worker_config.max_batch_size
+EV: <b>TYK_MDCB_SYNCWORKER_MAXBATCHSIZE</b><br />
+Type: `int`<br />
+
+The maximum amount of keys that we can fetch per batch. Default value: 1000 keys per batch.
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md]: Should be "maximum number" not "maximum amount"

File: tyk-docs/content/shared/mdcb-config.md [hunk 19]
```diff
@@ -285,99 +349,144 @@ Set it to true if you are using a separated analytic storage in the master gatew
...
+EV: <b>TYK_MDCB_SYNCWORKER_TIMEBETWEENBATCHES</b><br />
+Type: `int`<br />
+
+Specifies a cooldown time between batches in seconds. 0 / disabled by default.
+
+### sync_worker_config.max_workers
+EV: <b>TYK_MDCB_SYNCWORKER_MAXWORKERS</b><br />
+Type: `int`<br />
+
+To specify the maximum number of groups that can be synced at the same time without affecting the operation of MDCB. 1000 by default. It's recommendable to modify this value only if you need more org syncing at the same time.
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md]: Reword to:

"Specify the maximum number of groups that can be synced at the same time without affecting the operation of MDCB. 1000 by default. We recommend that you only modify this value if you need more org syncing at the same time."

File: tyk-docs/content/shared/mdcb-config.md [hunk 20]
```diff
@@ -285,99 +349,144 @@ Set it to true if you are using a separated analytic storage in the master gatew
...
+
+### sync_worker_config.warmup_time
+EV: <b>TYK_MDCB_SYNCWORKER_WARMUPTIME</b><br />
+Type: `int`<br />
+
+Specifies the time to wait before start syncing all the synchronization. It's especially useful to let the edge nodes synchronize APIs and Policies first before sync the other resources. Defaults to 2 seconds.
```
Comment (andyo-tyk) [inline tyk-docs/content/shared/mdcb-config.md]: Is "edge node" a "worker gateway"? For consistency in this document we should use "worker".
Is this option used to hold off the sync of resources other than APIs and Policies? So it's used to ensure that APIs and Policies are fully synched first? Why would we need to use this, does it not synchronise resources sequentially?

Perhaps reword to:

"The worker gateways synchronize APIs and Policies before other resources. Use this option to specify the time the worker should wait before starting the synchronization of those other resources to ensure that the critical sync (APIs and policies) has completed. Defaults to 2 seconds."

File: tyk-docs/content/shared/mdcb-config.md (hunk 2 above)
Comment (andyo-tyk) [reply tyk-docs/content/shared/mdcb-config.md:164]: OK, cool. Just the typo then, please. :-)

Is it obvious how TYK_MDCB_ANALYTICSCONFIG_CONNECTIONSTRING works with TYK_MDCB_ANALYTICSCONFIG_MONGOURL, both of which claim to define the connection string?

File: tyk-docs/content/shared/mdcb-config.md (hunk 1 above)
Comment (andyo-tyk) [reply tyk-docs/content/shared/mdcb-config.md:452]: Agreed to leave this and other wording as it is because we intend to upgrade all the MDCB documentation in one go.

File: tyk-docs/content/shared/mdcb-config.md (hunk 2 above)
Comment (andyo-tyk) [reply tyk-docs/content/shared/mdcb-config.md:164]: Agreed to leave this and other wording as it is because we intend to upgrade all the MDCB documentation in one go.

**Review by andyo-tyk — APPROVED** [review]

**Comment by sredxny** [comment]
@marksou we are ready to go! Please let us know when its published ;)

**Comment by marksou** [comment]
@sredxny @andyo-tyk now merged into v4 docs
